src/cart/src/cart/crt_context.c

```diff
@@ -1293,37 +1261,89 @@ crt_progress(crt_context_t crt_ctx, int64_t timeout,
 			hg_timeout = timeout;
 	}
 
-	while (true) {
+	/**
+	 * Call progress once before processing timeouts in case
+	 * any replies are pending in the queue
+	 */
+	rc = crt_hg_progress(&ctx->cc_hg_ctx, 0);
+	if (unlikely(rc && rc != -DER_TIMEDOUT)) {
+		D_ERROR("crt_hg_progress failed with %d\n", rc);
+		return rc;
+	}
+
+	/** loop until callback returns non-null value */
+	while ((rc = cond_cb(arg)) == 0) {
 		crt_context_timeout_check(ctx);
 		crt_exec_progress_cb(ctx);
 
 		rc = crt_hg_progress(&ctx->cc_hg_ctx, hg_timeout);
-		if (rc && rc != -DER_TIMEDOUT) {
+		if (unlikely(rc && rc != -DER_TIMEDOUT)) {
 			D_ERROR("crt_hg_progress failed with %d\n", rc);
-			D_GOTO(out, rc = 0);
+			return ret;
 		}
 
-		/** execute callback */
-		rc = cond_cb(arg);
-		if (rc > 0)
-			D_GOTO(out, rc = 0);
-		if (rc < 0)
-			D_GOTO(out, rc);
+		/** check for timeout */
+		if (timeout < 0)
+			continue;
 
-		/** check for timeout, if not infinite */
-		if (timeout > 0) {
-			now = d_timeus_secdiff(0);
-			if (now >= end) {
-				rc = -DER_TIMEDOUT;
+		now = d_timeus_secdiff(0);
+		if (timeout == 0 || now >= end) {
+			/** try callback one last time just in case */
+			rc = cond_cb(arg);
+			if (unlikely(rc != 0))
 				break;
-			}
-			if (end - now > 1000 * 1000)
-				hg_timeout = 1000 * 1000;
-			else
-				hg_timeout = end - now;
+			return -DER_TIMEDOUT;
 		}
+
+		/** adjust timeout */
+		if (end - now > 1000 * 1000)
+			hg_timeout = 1000 * 1000;
+		else
+			hg_timeout = end - now;
 	}
-out:
+
+	if (rc > 0)
+		rc = 0;
+
+	return rc;
+}
+
+int
+crt_progress(crt_context_t crt_ctx, int64_t timeout)
+{
+	struct crt_context	*ctx;
+	int			 rc = 0;
+
+	/** validate input parameters */
+	if (unlikely(crt_ctx == CRT_CONTEXT_NULL)) {
+		D_ERROR("invalid parameter (NULL crt_ctx).\n");
+		return -DER_INVAL;
+	}
+
+	ctx = crt_ctx;
+
+	/**
+	 * call progress once w/o any timeout before processing timed out
+	 * requests in case any replies are pending in the queue
+	 */
+	rc = crt_hg_progress(&ctx->cc_hg_ctx, 0);
+	if (unlikely(rc && rc != -DER_TIMEDOUT))
+		D_ERROR("crt_hg_progress failed, rc: %d.\n", rc);
+
+	/**
+	 * process timeout and progress callback after this initial call to
+	 * progress
+	 */
+	crt_context_timeout_check(ctx);
+	crt_exec_progress_cb(ctx);
+
+	if (timeout != 0 && (rc == 0 || rc == -DER_TIMEDOUT)) {
+		/** call progress once again with the real timeout */
+		rc = crt_hg_progress(&ctx->cc_hg_ctx, timeout);
```

Comment:
In this case, if  the previous call of crt_hg_progress() processed 256 events, we will call it again?

Reply:
yes, but w/o waiting. We were previously calling it with UINT_MAX. Do you think that this is a problem?